src/currentprocess/terminalsource.rs

```diff
@@ -0,0 +1,231 @@
+}
+
+// -------------- stdout -------------------------------
+
+/// Stand-in for std::io::stdout.
+pub trait StdoutTerminalSource {
+    fn stdout(&self) -> Box<dyn Terminal>;
+}
+
+// -------------- stderr -------------------------------
+
+/// Stand-in for std::io::stderr.
+pub trait StderrTerminalSource {
+    fn stderr(&self) -> Box<dyn Terminal>;
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum Attr {
+    Bold,
+    ForegroundColor(Color),
+}
+
+/// Disable all terminal controls on non-tty's.
+/// A wrapper around a `StandardStream` that provides a `Writer` implementation.
+struct StandardStreamTerminal {
+    stream: StandardStream,
+    color: ColorSpec,
+}
+
+impl Terminal for StandardStreamTerminal {
+    fn fg(&mut self, color: Color) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.color.set_fg(Some(color));
+        self.stream.set_color(&self.color)
+    }
+
+    fn bg(&mut self, color: Color) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.color.set_bg(Some(color));
+        self.stream.set_color(&self.color)
+    }
+
+    fn attr(&mut self, attr: Attr) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        match attr {
+            Attr::Bold => self.color.set_bold(true),
+            Attr::ForegroundColor(color) => self.color.set_fg(Some(color)),
+        };
+        self.stream.set_color(&self.color)
+    }
+
+    fn reset(&mut self) -> io::Result<()> {
+        if !self.stream.supports_color() {
+            return Ok(());
+        }
+        self.stream.reset()
+    }
+
+    fn carriage_return(&mut self) -> io::Result<()> {
+        let _ = self.stream.write(b"\r")?;
+        Ok(())
+    }
+}
+
+impl WriterLock for termcolor::StandardStreamLock<'_> {}
+
+impl Writer for StandardStreamTerminal {
+    fn lock(&self) -> Box<dyn WriterLock + '_> {
+        Box::new(self.stream.lock())
+    }
+}
+
+impl io::Write for StandardStreamTerminal {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.stream.write(buf)
+    }
+
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.stream.flush()
+    }
+}
+
+/// Returns a `StandardStreamTerminal` for stdout. If stdout is not a tty, then
+/// the returned `StandardStreamTerminal` will not support coloring.
+fn stdout() -> StandardStreamTerminal {
+    let choice = if process().is_stdout_a_tty() {
+        ColorChoice::Auto
+    } else {
+        ColorChoice::Never
+    };
+    StandardStreamTerminal {
+        stream: termcolor::StandardStream::stdout(choice),
+        color: ColorSpec::new(),
+    }
+}
+
+/// Returns a `StandardStreamTerminal` for stderr. If stderr is not a tty, then
+/// the returned `StandardStreamTerminal` will not support coloring.
+fn stderr() -> StandardStreamTerminal {
```

Comment:
Genuine question - do we want a terminal abstraction on stderr? If so then for my suggestion above about `.terminal()` we could add `.terminal()` to the Stdout / Stderr traits. 

So it would look like

`process.stderr().terminal()`

Reply:
I guess there is no good way to do that. Because the stderr is a writer and the terminal is also a writer, it does not make sense to implement a terminal for a writer.

Reply:
I think there are two separate discussions here that perhaps we're not connecting on.

One is what termcolor requires: we agree that it has its own definition of stdout/stderr etc and can't layer on a provided struct/trait.

The other is what we offer *within* rustup. 
A) Do we supply a termcolor terminal (even wrapped by a struct) to code that is doing low-level operations or
B) do we supply a low level struct with less features to such code, and a termcolor struct (wrapped sure) to higher level code

I think B is better. The lock() feature will prevent cross talk between these layers (and we should also do more like making progress bars run on stderr, but e.g. JSON output on stdout).